cpp/include/cudf/detail/gather.cuh

```diff
@@ -46,58 +62,83 @@ struct bounds_checker {
 };
 
 
+/**----------------------------------------------------------------------------*
+ * @brief For each mask, assigns each bit to a value obtained by calling the
+ *        validity selector.
+ *
+ * @note If any mask is `nullptr`, that mask will be ignored.
+ *
+ * @param get_validity    Function used to obtain validity for each bit.
+ *                        Signature: `bool(size_type mask_idx, size_type bit_idx)`
+ * @param masks           Masks for which bits will be obtained and assigned.
+ * @param mask_count      The number of `masks`.
+ * @param mask_num_bits   The number of bits to assign for each mask. If this
+ *                        number is smaller than the total number of bits, the
+ *                        remaining bits will be set to `0` / `false`.
+ * @param validity_counts Used to obtain the total number of valid bits for each
+ *                        mask.
+ *
+ * @tparam ValiditySelector bool(size_type mask_idx, size_type bit_idx);
+ *---------------------------------------------------------------------------**/
+template <typename ValiditySelector, int32_t block_size>
+__global__ void select_bitmask_kernel(ValiditySelector get_validity,
+                                      bitmask_type* masks[],
+                                      size_type mask_count,
+                                      size_type mask_num_bits,
+                                      size_type* valid_counts)
+{
+  for (size_type mask_idx = 0; mask_idx < mask_count; mask_idx++) {
+    auto const mask = masks[mask_idx];
+    if (mask == nullptr) {
+      continue;
+    }
+
+    auto lane_offset = blockIdx.x * blockDim.x;
```

Comment:
This naming is a bit misleading. "lane" refers to a thread's position within a warp, so valid values are `[0,32)`. Something like `block_offset` would be more appropriate.